src/dotnet/UI.Blazor/Services/HistoryUI/HistoryUI.cs

```diff
@@ -0,0 +1,228 @@
+            HistoryMove move;
+            // TODO: in .NET 7 NavigationManager provides access to state property of the history API.
+            // Rework history position tracking with it.
+            var readState = await GetStateAsync();
+            if (readState == null) {
+                // State is null, apparently it's an internal navigation to next location happened.
+                // Lets store this in the state.
+                var state = new State { Index = previousState.Index + 1 };
+                await SetStateAsync(state);
+                _state = state;
+                move = HistoryMove.Navigate;
+            }
+            else {
+                move = readState.Index < previousState.Index
+                    ? HistoryMove.Backward
+                    : readState.Index > previousState.Index
+                        ? HistoryMove.Forward
+                        : throw StandardError.Constraint("Invalid state index for history move.");
+                _state = readState;
+            }
+
+            HistoryItem historyItem;
+            if (move == HistoryMove.Navigate) {
+                if (_pendingHistoryItem != null) {
+                    if (_pendingHistoryItem.HistoryIndex != _historyIndex)
+                        throw StandardError.Constraint("PendingHistoryItem is not consistent.");
+                    historyItem = _pendingHistoryItem.HistoryItem;
+                    _pendingHistoryItem = null;
+                }
+                else
+                    historyItem = new HistoryItem(location);
+                _historyIndex++;
+                if (_history.Count < _historyIndex)
+                    throw StandardError.Constraint("History is not consistent.");
+                if (_history.Count > _historyIndex)
+                    _history.RemoveRange(_historyIndex, _history.Count - _historyIndex);
+                _history.Add(historyItem);
+            }
+            else if (move == HistoryMove.Forward) {
+                _historyIndex = readState!.Index;
+                if (_history.Count < _historyIndex)
+                    throw StandardError.Constraint("History does not contain forward item.");
+                historyItem = _history[_historyIndex];
+            } else {
+                _historyIndex = readState!.Index;
+                if (_history.Count < _historyIndex + 1)
+                    throw StandardError.Constraint("History does not contain backward item.");
+                historyItem = _history[_historyIndex + 1];
+            }
+
+            if (move != HistoryMove.Backward)
+                historyItem.OnForwardAction?.Invoke();
+            else
+                historyItem.OnBackAction?.Invoke();
+
+            var eventsArgs = new AfterLocationChangedHandledEventsArgs {
+                Location = location,
+                HasLocationChanged = OrdinalEquals(location, previousLocation),
+                IsBackward = move == HistoryMove.Backward
```

Comment:
this assumes, that for both `Navigate` and `Forward` IsBackward will be false. Is it significant?

Reply:
imho, not used code == not tested at all == legacy == buggy code

Reply:
Removed no longer used code